reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,84 +83,124 @@ object StreamSubscriber {
     case object OnFinalize extends Input
-    case class OnDequeue(response: Deferred[F, Out]) extends Input
+    case object OnDequeue extends Input
 
     sealed trait State
     case object Uninitialized extends State
     case class Idle(sub: Subscription) extends State
-    case class RequestBeforeSubscription(req: Deferred[F, Out]) extends State
-    case class WaitingOnUpstream(sub: Subscription, elementRequest: Deferred[F, Out]) extends State
+    case class Receiving(sub: Subscription) extends State
+    case object RequestBeforeSubscription extends State
     case object UpstreamCompletion extends State
     case object DownstreamCancellation extends State
     case class UpstreamError(err: Throwable) extends State
 
     def step(in: Input): State => (State, F[Unit]) =
       in match {
         case OnSubscribe(s) => {
-          case RequestBeforeSubscription(req) => WaitingOnUpstream(s, req) -> F.delay(s.request(1))
-          case Uninitialized                  => Idle(s) -> F.unit
+          case RequestBeforeSubscription =>
+            Receiving(s) -> F.delay(s.request(maxDemand))
+          case Uninitialized =>
+            Idle(s) -> F.unit
+          case Idle(prev) =>
+            Idle(prev) -> F.delay(s.cancel)
+          case Receiving(prev) =>
+            Receiving(prev) -> F.delay(s.cancel)
           case o =>
             val err = new Error(s"received subscription in invalid state [$o]")
             o -> (F.delay(s.cancel) >> F.raiseError(err))
         }
         case OnNext(a) => {
-          case WaitingOnUpstream(s, r) => Idle(s) -> r.complete(a.some.asRight)
-          case DownstreamCancellation  => DownstreamCancellation -> F.unit
-          case o                       => o -> F.raiseError(new Error(s"received record [$a] in invalid state [$o]"))
+          case Receiving(s) =>
+            Receiving(s) -> q.enqueue1(a.some.asRight)
+          case Idle(s) =>
+            Receiving(s) -> q.enqueue1(a.some.asRight)
+          case DownstreamCancellation =>
+            DownstreamCancellation -> F.unit
+          case o =>
+            o -> F.raiseError(new Error(s"received record [$a] in invalid state [$o]"))
         }
         case OnComplete => {
-          case WaitingOnUpstream(_, r) => UpstreamCompletion -> r.complete(None.asRight)
-          case _                       => UpstreamCompletion -> F.unit
+          case Receiving(_) =>
+            UpstreamCompletion -> q.enqueue1(None.asRight)
+          case Idle(_) =>
+            UpstreamCompletion -> q.enqueue1(None.asRight)
+          case _ =>
+            UpstreamCompletion -> F.unit
         }
         case OnError(e) => {
-          case WaitingOnUpstream(_, r) => UpstreamError(e) -> r.complete(e.asLeft)
-          case _                       => UpstreamError(e) -> F.unit
+          case Receiving(_) =>
+            UpstreamError(e) -> (q.enqueue1(e.asLeft) >> q.enqueue1(None.asRight))
+          case _ =>
```

Comment:
We could also be in `Idle` here (according to the spec rule 2.10), which we should handle like the Receiving case.